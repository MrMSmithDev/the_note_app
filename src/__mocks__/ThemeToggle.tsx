export default function ThemeToggle() {
  return <div data-testid="mock-theme-toggle"></div>;
}
