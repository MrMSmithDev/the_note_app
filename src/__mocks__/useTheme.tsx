export const mockTheme = 'light';
export const mockToggleTheme = jest.fn(() => {});

const useTheme = jest.fn(() => [mockTheme, mockToggleTheme]);

export default useTheme;
