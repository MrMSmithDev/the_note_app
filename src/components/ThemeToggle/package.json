{
    "name": "theme-toggle",
    "version": "1.0.0",
    "main": "ThemeToggle.tsx",
    "dependencies": {
      "react": "^19.0.0"
    }
}