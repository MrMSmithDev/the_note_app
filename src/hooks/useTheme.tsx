import { useState } from 'react';

type ThemeType = 'light' | 'dark';

const useTheme = (): [string, () => void] => {
  const [theme, setTheme] = useState<ThemeType>(() => {
    if (typeof window !== 'undefined') {
      return (localStorage.getItem('theme') ||
        (window.matchMedia('(prefers-color-scheme: dark)').matches
          ? 'dark'
          : 'light')) as ThemeType;
    }
    return 'light';
  });

  function toggleTheme() {
    const newTheme = theme === 'light' ? 'dark' : 'light';

    document.documentElement.classList.remove(theme);
    document.documentElement.classList.add(newTheme);

    setTheme(newTheme);
    localStorage.setItem('theme', newTheme);
  }

  return [theme, toggleTheme];
};

export default useTheme;
