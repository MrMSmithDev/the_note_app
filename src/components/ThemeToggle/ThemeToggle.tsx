import React, { useEffect, useState } from 'react';
import DarkModeIcon from '@components/icons/DarkModeIcon';
import LightModeIcon from '@components/icons/LightModeIcon';
import useTheme from '@hooks/useTheme';

const ThemeToggle: React.FC = () => {
  const [theme, themeToggle, themeLoading] = useTheme();
  const [icon, setIcon] = useState<string>('light');

  useEffect(() => {
    theme === 'light' ? setIcon('light') : setIcon('dark');
  }, [theme, themeLoading]); // force rerender on load if previously set to dark

  return (
    <div className="flex items-center ml-3">
      <button onClick={themeToggle} className="cursor-pointer">
        {icon === 'light' ? (
          <DarkModeIcon className="h-3/4 w-3/4" />
        ) : (
          <LightModeIcon className="h-3/4 w-3/4" />
        )}
      </button>
    </div>
  );
};

export default ThemeToggle;
