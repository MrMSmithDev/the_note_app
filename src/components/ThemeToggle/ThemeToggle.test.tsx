import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import ThemeToggle from './ThemeToggle';
import useTheme from '@hooks/useTheme';

jest.mock('@hooks/useTheme');

describe('ThemeToggle component', () => {
  beforeEach(() => {
    jest.resetAllMocks(); // Reset mocks before each test
  });

  it('renders correctly to match snapshot', () => {
    (useTheme as jest.Mock).mockReturnValue(['light', jest.fn()]);

    const { asFragment } = render(<ThemeToggle />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the opposite icon based on theme', async () => {
    (useTheme as jest.Mock).mockReturnValue(['light', jest.fn()]);

    render(<ThemeToggle />);
    expect(screen.getByTestId('dark-mode-icon')).toBeInTheDocument();

    (useTheme as jest.Mock).mockReturnValueOnce(['dark', jest.fn()]);

    render(<ThemeToggle />);

    expect(screen.getByTestId('light-mode-icon')).toBeInTheDocument();
  });

  it('calls themeToggle when the button is clicked', () => {
    const mockToggleFunc = jest.fn();
    (useTheme as jest.Mock).mockReturnValueOnce(['light', mockToggleFunc]);

    render(<ThemeToggle />);

    fireEvent.click(screen.getByRole('button')); // Simulate button click

    expect(mockToggleFunc).toHaveBeenCalled();
  });
});
