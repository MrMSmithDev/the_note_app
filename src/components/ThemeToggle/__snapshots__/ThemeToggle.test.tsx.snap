// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeToggle component renders correctly to match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center ml-3"
  >
    <button
      class="cursor-pointer"
    >
      <svg
        class="h-3/4 w-3/4"
        data-testid="dark-mode-icon"
        fill="#00035B"
        height="34px"
        viewBox="0 -960 960 960"
        width="34px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M480-120q-150 0-255-105T120-480q0-150 105-255t255-105q14 0 27.5 1t26.5 3q-41 29-65.5 75.5T444-660q0 90 63 153t153 63q55 0 101-24.5t75-65.5q2 13 3 26.5t1 27.5q0 150-105 255T480-120Z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
